Netpbm Editor

SilasBerger opened the pull request:
Persistable Netpbm editor (+ new document type) and live viewer.

lebalz commented on src/components/documents/NetpbmEditor/index.tsx:
```diff
@@ -0,0 +1,151 @@
+        </span>
+    );
+
+    const hasErrorsOrWarnings = () => displayedErrors.length > 0 || displayedWarnings.length > 0;
+    const hasWarnings = () => displayedWarnings.length > 0;
+    const hasErrors = () => displayedErrors.length > 0;
```
[perf]: use `React.useMemo`: it will recompute only when it's dependencies change

```suggestion
    const { hasErrorsOrWarnings, hasWarnings, hasErrors } = React.useMemo(() => {
        return {
            hasErrorsOrWarnings: displayedErrors.length > 0 || displayedWarnings.length > 0,
            hasWarnings: displayedWarnings.length > 0,
            hasErrors: displayedErrors.length > 0
        };
    }, [displayedErrors, displayedWarnings]);
```

SilasBerger replied:
TIL – cool feature 😁

lebalz commented on src/components/documents/NetpbmEditor/types.tsx:
Should this file be named `types.d.ts` as it contains types only?

lebalz reviewed (approved):
Really cool addition, fancy! 🤩 i think this is ready to merge - my comments can be either addressed as part of this pr or i'm fine to work 'em up in a follow-up pr...
Great Work! Thanks 🚀🥳🥳

SilasBerger commented on src/components/documents/NetpbmEditor/parser/parser.tsx:
```diff
@@ -0,0 +1,76 @@
+
+export const SUPPORTED_FORMATS = ['P1', 'P2', 'P3'];
+
+export const PATTERN_P1 = /(?<format>P1)\s+(?<width>\d+)\s+(?<height>\d+)\s+(?<raster>[\d\D\s]*)/;
```
Yeah, I noticed that too. The problem is that JS can still parse `'1wait'` as `1` (go figure 😄). I initially considered running each byte character through a regex, but was worried about performance. I tried it now though, and performance doesn't seem to be an issue. These values are now no longer allowed 😉.

SilasBerger commented on src/components/documents/NetpbmEditor/parser/parser.tsx:
```diff
@@ -0,0 +1,76 @@
+
+    const lines = sanitizedData.split('\n');
+    if (!SUPPORTED_FORMATS.includes(lines[0].trim())) {
+        errors.push(
+            <span>
+                Unbekanntes Format auf der ersten Zeile: <code>{lines[0].trim()}</code>. Unterstützte Formate:{' '}
+                <code>{SUPPORTED_FORMATS.join(', ')}</code>.
+            </span>
+        );
```
I agree, it's not quite the clean solution. I'll leave it as-is for now, but I will probably implement your suggestion in a follow-up at some point :).

SilasBerger commented on src/components/documents/NetpbmEditor/parser/rasterParser.tsx:
```diff
@@ -0,0 +1,180 @@
+                    );
+                }
+
+                pixels[row * width * 4 + col * 4] = bitValue * 255; // red
+                pixels[row * width * 4 + col * 4 + 1] = bitValue * 255; // green
+                pixels[row * width * 4 + col * 4 + 2] = bitValue * 255; // blue
+                pixels[row * width * 4 + col * 4 + 3] = 255; // alpha
```
Nice, I like that too! Let's keep this in mind in case performance becomes an issue.

SilasBerger commented:
> Really cool addition, fancy! 🤩 i think this is ready to merge - my comments can be either addressed as part of this pr or i'm fine to work 'em up in a follow-up pr... Great Work! Thanks 🚀🥳🥳

Thank you very much, and thanks for the review 😊 . I addressed most of your comments now and I'm going to merge the PR, with possible follow-ups.
